Add a new Link component

dunglas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  |no <!-- don't forget updating src/**/CHANGELOG.md files -->
| BC breaks?    | no
| Deprecations? | no <!-- don't forget updating UPGRADE-*.md files -->
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | todo

This a proposal to extract HTTP preloading features introduced in #21478 in their own component. 

There are some good reasons to do it:

* HTTP preloading is not (only) about assets: this standalone component could be very useful to replace resources embedding in APIs by HTTP/2 pushes like described in [this article](https://evertpot.com/rest-embedding-hal-http2/) by @evert. In such case, there is no reason to carry the whole asset component for an API.
* There is no dependency nor relation at all between the code of the asset compnent and the one I've added for Preloading features
* It makes the code cleaner (no more optional dependency in the `Asset` Twig extension)

This component would also better fit in HttpFoundation than in Asset. But there is no dependency between it and HttpFoundation and it can easily be used with PSR-7 too, so IMO it better belongs in a standalone component.

Btw, ~~~I plan to add support for prefetching to this component. Except a PR soon.~~~ Prefetching and prerendering support added in this PR.

ping @symfony/deciders

dunglas committed:
Add a new Preload component

dunglas commented:
A better name would probably be "Link Component": https://html.spec.whatwg.org/multipage/semantics.html#links

nicolas-grekas commented:
Added to the 3.3 milestone because this is only about new features already planned for 3.3

stof commented:
Not changing the service definition for `assets.preload_manager` looks suspicious to me (and btw, the name should be changed)

dunglas committed:
Rename the component Link and add generic Link support

dunglas commented:
I've renamed the component Link, added support for generic "Link" headers (preloading and resource hints) and introduced new Twig helpers for `dns-prefetch`, `preconnect`, `prefetch` and `prerender`.

This new component now support all types of prefetching and is smart enough to manage generic Link headers (for instance, when it will be available https://github.com/api-platform/core/blob/master/src/Hydra/EventListener/AddLinkHeaderListener.php in API Platform may be removed).

@stof comment also addressed.

tadcka commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -39,6 +39,7 @@
         "symfony/console": "~3.3",
         "symfony/css-selector": "~2.8|~3.0",
         "symfony/dom-crawler": "~2.8|~3.0",
+        "symfony/links": "~3.3",
```
symfony/link

andersonamuller commented on src/Symfony/Component/Link/composer.json:
```diff
@@ -0,0 +1,39 @@
+        "symfony/http-kernel": "^2.8 || ^3.0"
+    },
+    "autoload": {
+        "psr-4": { "Symfony\\Component\\Preload\\": "" },
```
Symfony\\Component\\Link\\

dunglas commented:
Tests should be green after the merge.

xabbuh commented on src/Symfony/Bridge/Twig/Tests/Extension/LinkExtensionTest.php:
```diff
@@ -0,0 +1,83 @@
+{
+    protected function setUp()
+    {
+        if (!class_exists(LinkManager::class)) {
```
I would add `symfony/link` to the `require-dev` section of the `composer.json` file instead.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -208,6 +208,10 @@ public function load(array $configs, ContainerBuilder $container)
             $this->registerPropertyInfoConfiguration($config['property_info'], $container, $loader);
         }
 
+        if ($this->isConfigEnabled($container, $config['links'])) {
```
I think we should test for the Link component to be present first and throw a meaningful exception if not present.

dunglas replied:
Link should not be a mandatory component IMO. You mean throwing if enabled is set to `true` but the component not installed?

xabbuh replied:
yes, we did that recently for other optional dependencies too

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/links.xml:
```diff
@@ -0,0 +1,18 @@
+
+        <service id="links.link_manager" class="Symfony\Component\Link\LinkManager" public="false" />
+
+        <service id="links.link_listener" class="Symfony\Component\Link\EventListener\LinkListener">
```
Can be private?

dunglas replied:
IIRC, event listeners can't

xabbuh replied:
since #20953 they can

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:
```diff
@@ -62,6 +63,10 @@
     </xsd:complexType>
 
+    <xsd:complexType name="links">
+        <xsd:attribute name="enabled" type="xsd:boolean" />
```
Does it make sense to set this to `xsd:string` to allow using a DI parameter in the config?

dunglas replied:
All others similar parameters use `xsd:boolean`. If we want to change this behavior, we should do it in another PR for all `enabled` parameters.

nicolas-grekas replied:
maybe something to "fix" by adding a new "boolean" type, borrowed from the XSD of the DI XmlFileLoader?

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -238,6 +238,9 @@ protected static function getBundleDefaultConfig()
                 'throw' => true,
             ),
+            'links' => array(
+                'enabled' => !class_exists(FullStack::class),
```
What about adding the component as a dev dependency to the `composer.json` file instead?

dunglas replied:
It is, it's not related.

xabbuh replied:
sorry, I misread the code

xabbuh commented on src/Symfony/Component/Link/LinkManager.php:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class LinkManager implements LinkManagerInterface
```
`final`?

dunglas replied:
👍 on my side but I think it's against our CS (maybe `@final`). ping @nicolas-grekas

xabbuh commented on src/Symfony/Component/Link/README.md:
```diff
@@ -0,0 +1,14 @@
+Link Component
+==============
+
+The Link component manages link between resources. It is particularly useful advise clients
```
[...] manages links between [...]

xabbuh replied:
[...] useful to advise [...]

xabbuh commented on src/Symfony/Component/Link/composer.json:
```diff
@@ -0,0 +1,39 @@
+        "symfony/http-kernel": ""
+    },
+    "require-dev": {
+        "symfony/http-kernel": "^2.8 || ^3.0"
```
The EventDispatcher and HttpFoundation components are used by the `LinkListenerTest` class too.

teohhanhui commented:
Please, call this "HttpLink".

dunglas commented:
@teohhanhui it's not really related to HTTP (only the listener is). It can also handle HTML links and in general links in any other protocol.

dunglas commented:
@xabbuh's comments fixed

teohhanhui commented:
I believe it's defined in [RFC 5988 - Web Linking](https://tools.ietf.org/html/rfc5988). Perhaps it can be named WebLink or ResourceLink.

dunglas commented:
I like `WebLink`.

andersonamuller commented:
`WebLink` is fine, but then the `buildValues` method on `LinkManagerInterface` should not mention `HTTP headers`

dunglas commented:
This method should be renamed `buildHeaderValue()` (obviously, it's not the same format for HTML links).

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/links.xml:
```diff
@@ -0,0 +1,18 @@
+
+    <services>
+
+        <service id="links.link_manager" class="Symfony\Component\Link\LinkManager" public="false" />
```
this file is missing autowiring aliases

nicolas-grekas commented on src/Symfony/Component/Link/LinkManagerInterface.php:
```diff
@@ -0,0 +1,45 @@
+     * Adds an element to the list of resources to preload.
+     *
+     * @param string $uri        The relation URI
+     * @param string $rel        The relation type (e.g. "preload", "prefetch", "prerender" or "dns-prefetch")
```
worth adding constants for the existing types?

evert replied:
There are hundreds, and it's extendable with uris

nicolas-grekas replied:
some are special, at least here: https://github.com/symfony/symfony/pull/22273/files#diff-55d869d95fa50fe827384c83ea961248R33

dunglas replied:
Does it really worth it? Those names make sense in the Twig extension because they are useful for assets. In the component, as pointed out by @evert they can be a lot values.

nicolas-grekas replied:
there are hundreds of them, but only a very few are standardized for technical use (eg only 4 in https://www.w3.org/TR/resource-hints/)
So to me these may deserve a const to help discoverability.

nicolas-grekas commented on src/Symfony/Component/Link/README.md:
```diff
@@ -0,0 +1,14 @@
+==============
+
+The Link component manages links between resources. It is particularly useful to advise clients
+to preload and prefetch documents through HTTP and HTTP/2 pushes.
```
if the component follows some sort of standard, it could be worth mentioning it here

nicolas-grekas reviewed (commented):
Could "Link" be a too generic name? could we need a more scoped name? HtmlLink?
Just wondering - keep "Link" might be the best :)

Simperfit commented on src/Symfony/Bridge/Twig/Extension/LinkExtension.php:
```diff
@@ -0,0 +1,135 @@
+use Symfony\Component\Link\LinkManagerInterface;
+
+/**
+ * Twig extension for the Symfony Preload component.
```
`WebLink` is a good name ;)

dunglas committed:
Rename to WebLink

mnapoli commented:
Is there any relation with PSR-13 about links? Would it make sense to implement it?

For reference:

- the GitHub package: https://github.com/php-fig/link
- the standard: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-13-links.md

I am not familiar at all with the standard so I may be completely off.

fabpot commented on src/Symfony/Component/WebLink/composer.json:
```diff
@@ -0,0 +1,41 @@
+{
+    "name": "symfony/web-link",
```
I think I would prefer `weblink`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -39,6 +39,7 @@
         "symfony/console": "~3.3",
         "symfony/css-selector": "~2.8|~3.0",
         "symfony/dom-crawler": "~2.8|~3.0",
+        "symfony/link": "~3.3",
```
web-link or weblink

fabpot commented on src/Symfony/Bundle/TwigBundle/composer.json:
```diff
@@ -33,6 +33,7 @@
         "symfony/routing": "~2.8|~3.0",
         "symfony/templating": "~2.8|~3.0",
         "symfony/yaml": "~2.8|~3.0",
+        "symfony/link": "~3.3",
```
weblink or web-link

dunglas commented:
@mnapoli I had totally forgotten this PSR, but it's related. We may use https://github.com/php-fig/link-util instead of the primitive types. Are you aware of other PSR-13 implementations?

mnapoli commented:
@dunglas not really, pinging @Crell who may have some insight. On Packagist there are not a lot of dependent packages: https://packagist.org/packages/psr/link/dependents

dunglas commented:
So, about PSR-13, the manager can be transformed to a serializer (according to the PSR terminology) and use https://github.com/php-fig/link-util/blob/master/src/GenericLinkProvider.php instead of a simple array.

It's a minor change and will improve operability with other libraries using this PSR (Drupal?). WDYT @symfony/deciders?

nicolas-grekas commented:
Let's use PSR-13 then!

dunglas committed:
Use PSR-13

dunglas commented:
PSR-13 support added (it deserves a review because of how I deal with immutability).

dunglas commented:
Failures not related or will be fixed by the merge.

teohhanhui commented on composer.json:
```diff
@@ -23,6 +23,7 @@
         "psr/container": "^1.0",
         "psr/log": "~1.0",
         "psr/simple-cache": "^1.0",
+        "fig/link-util": "^1.0",
```
I think we should require `psr/link` directly too? Because `fig/link-util` is just an implementation, even though it does add a transitive dependency on `psr/link`.

nicolas-grekas replied:
any `psr/link-implementation` to add also in provide section?

teohhanhui replied:
How easy it is to define a virtual package (basically just list it in `provide` - that's why it's called virtual):
https://devedge.wordpress.com/2014/09/27/composer-and-virtual-packages/

And some counter points:
https://php-and-symfony.matthiasnoback.nl/2014/10/composer-provide-and-dependency-inversion/

dunglas replied:
We use the implementations here: https://github.com/symfony/symfony/pull/22273/files#diff-78128eb201bbbb556131b3ae4bac192bR29.
We may require only a `psr/link-implementation` instead of `fig/link-util` but it will hurt the DX (the constructor argument is optional). I'm to let it as is because you can already use any implementation but it also works out of the box.

teohhanhui commented on src/Symfony/Component/WebLink/composer.json:
```diff
@@ -0,0 +1,42 @@
+{
+    "name": "symfony/weblink",
```
This is just very confusing and inconsistent. We either rename the component `Weblink` (bad idea), or we should stick to `symfony/web-link`.

dunglas replied:
We indeed use a caret in other components (`event-dispatcher`, `http-foundation`), but as @fabpot wants.

teohhanhui replied:
@fabpot seemed ok with `symfony/web-link`: https://github.com/symfony/symfony/pull/22273#pullrequestreview-31173146

dunglas replied:
@teohhanhui it was called `web-link` and he asked for a change :) https://github.com/symfony/symfony/pull/22273#discussion_r110030764
I've no strong opinion about that even if I prefer with the caret.

jvasseur replied:
Maybe we could rename the namespace to `Weblink` instead to keep consistency between package name and namespace.

fabpot replied:
What about going back to "Link":

 * The PSR is named Link, not Weblink;
 * Short names are better;
 * No dash (that helps :));
 * Do we think that we will have a non-web link component at some point? (or do we think that Link can be confusing?)

WDYT? If everybody think that web-link is better, that's of course an option as well.

dunglas replied:
The (new) problem with Link is that it's not an implementation of the PSR-13, it just use it. It maybe confusing :)

`symfony/link` and `symfony/web-link` are fine to me. It's time to start a Twitter poll?

teohhanhui replied:
I think Link is too generic. This component deals with Web Linking. It's a very specific concept. When you say link, people usually think of URI, which is not what this is about.

nicolas-grekas replied:
the confusion is non existent: that's not the same vendor, people do not know about all PSRs
the generic name is also not an issue: the fig already use that name, so it wasn't too generic for them, and again we have the "symfony" prefix, which really suggests "web" to many.

I'm :+1: for going back to just "Link"

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -18,7 +18,7 @@
 use Symfony\Component\Config\Definition\ConfigurationInterface;
 use Symfony\Component\Form\Form;
-use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\WebLink\WebLinkManagerInterface;
```
alpha order

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -29,6 +29,7 @@
 use Symfony\Component\Finder\Finder;
 use Symfony\Component\HttpKernel\DependencyInjection\Extension;
 use Symfony\Component\Config\FileLocator;
+use Symfony\Component\WebLink\WebLinkManagerInterface;
```
alpha order

nicolas-grekas commented on src/Symfony/Component/WebLink/WebLinkManagerInterface.php:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+interface WebLinkManagerInterface
```
would it be possible to have only the PSR interfaces and not add any new in the component? that'd play well with the goal of the psr

dunglas replied:
Yes, but we still need the manager because we need a "mutable" service to use with event listeners.

nicolas-grekas replied:
OK, let's keep it, that allows swapping the implem of the service

nicolas-grekas commented on src/Symfony/Component/WebLink/EventListener/AddLinkHeaderListener.php:
```diff
@@ -9,26 +9,30 @@
-use Symfony\Component\Asset\Preload\PreloadManager;
-use Symfony\Component\Asset\Preload\PreloadManagerInterface;
+use Symfony\Component\WebLink\HttpHeaderSerializer;
+use Symfony\Component\WebLink\WebLinkManagerInterface;
```
alpha order

nicolas-grekas commented on src/Symfony/Component/WebLink/README.md:
```diff
@@ -0,0 +1,18 @@
+WebLink Component
+=================
+
+The Link component manages links between resources. It is particularly useful to advise clients
```
The WebLink

sstok commented on src/Symfony/Component/WebLink/Tests/WebLinkManagerTest.php:
```diff
@@ -0,0 +1,52 @@
+        $this->assertSame(array($link1, $link2), array_values($this->manager->getLinkProvider()->getLinks()));
+    }
+
+    public function testClear() {
```
`{` should be on new line 😛 is fabbot on a break?

dunglas committed:
Fix naming and CS

dunglas committed:
Remove the dependency to fig/link-util

fabpot commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -79,7 +81,8 @@
         "symfony/property-info": "For using the property_info service",
-        "symfony/process": "For using the server:run, server:start, server:stop, and server:status commands"
+        "symfony/process": "For using the server:run, server:start, server:stop, and server:status commands",
+        "symfony/web-links": "For using web links features such as preloading, prefetching or prerendering"
```
`web-link`

dunglas committed:
Inline the GenericLinkProvider service

fabpot commented on src/Symfony/Bridge/Twig/Extension/WebLinkExtension.php:
```diff
@@ -0,0 +1,141 @@
+    public function getName()
+    {
+        return 'web_link';
+    }
```
We don't need this method.

fabpot commented on src/Symfony/Bridge/Twig/Extension/WebLinkExtension.php:
```diff
@@ -0,0 +1,141 @@
+{
+    private $manager;
+
+    public function __construct(WebLinkManagerInterface $manager)
```
Instead of injecting this dep here, it would be better to create a Twig Runtime.

dunglas committed:
Fix composer.json

nicolas-grekas commented on src/Symfony/Bridge/Twig/Extension/WebLinkExtension.php:
```diff
@@ -0,0 +1,141 @@
+            $link = $link->withAttribute($key, $value);
+        }
+
+        $this->manager->add($link);
```
as discussed with @dunglas on Slack, this mean the manager is mutable, which conflist with the fact that it is registered as a service
using an attribute on the request might be the solution

dunglas committed:
Use a request attribute instead of a stateful service

dunglas commented on src/Symfony/Bridge/Twig/Extension/WebLinkExtension.php:
```diff
@@ -0,0 +1,148 @@
+            $link = $link->withAttribute($key, $value);
+        }
+
+        $linkProvider = $request->attributes->get('_links', new GenericLinkProvider());
```
Is this ok or should we always populate this key? The current implementation is more memory efficient.

teohhanhui replied:
Actually, I really think the ideal case would be for `Symfony\Component\HttpFoundation\Request` to implement `Psr\Link\EvolvableLinkProviderInterface`. That's what I understand after reading PSR-13.

But I understand that's not probably not possible considering the mutable nature of our `Request`. That said, I think using attributes for this is wrong.

dunglas replied:
> Actually, I really think the ideal case would be for Symfony\Component\HttpFoundation\Request to implement Psr\Link\EvolvableLinkProviderInterface.

I agree but as you pointed out, it's not possible.

> That said, I think using attributes for this is wrong.

Why?

dunglas commented:
Implementation revamped to get rid of the stateful service and use a request attribute instead.

dunglas committed:
Remove the getName method

nicolas-grekas commented on src/Symfony/Component/WebLink/composer.json:
```diff
@@ -0,0 +1,43 @@
+        "fig/link-util": "^1.0",
+        "symfony/event-dispatcher": "^2.8 || ^3.0",
+        "symfony/http-foundation": "^2.8 || ^3.0",
+        "symfony/http-kernel": "^2.8 || ^3.0"
```
the convention is `^2.8|^3.0`

nicolas-grekas commented on src/Symfony/Component/WebLink/EventListener/AddLinkHeaderListener.php:
```diff
@@ -37,12 +39,12 @@ public function onKernelResponse(FilterResponseEvent $event)
-            // Free memory
-            $this->preloadManager->clear();
+        $linkProvider = $event->getRequest()->attributes->get('_links');
+        if (!$linkProvider instanceof LinkProviderInterface || !($links = $linkProvider->getLinks())) {
```
no need for the brackets after the `||`

nicolas-grekas commented on composer.json:
```diff
@@ -18,9 +18,11 @@
     "require": {
         "php": ">=5.5.9",
         "doctrine/common": "~2.4",
+        "fig/link-util": "^1.0",
```
this dep is in require-dev in all components, but in require here, looks suspicious

dunglas replied:
We need this package in the Twig extension and the FrameworkBundle  (even at runtime). But the those packages have only a soft dependency to WebLink. In the context of the full stack, I think it makes sense to add it to have something working out of the box.

teohhanhui replied:
It's not `require`d by any of the packages in `replace`, so it should indeed remain in `require-dev`.

dunglas replied:
It is required if you use WebLink && (Twig Bridge || Framework Bundle). We cannot express such conditions with Composer, but the full stack framework install WebLink + Twig Bridge + Framework Bundle so it needs to have this package as dependency.

nicolas-grekas replied:
Still looks inconsistent to me. symfony/symfony should not behave differently than using standalone components...

dunglas replied:
`fig/link-util` added back as a dependency of the WebLink component. It fixes the inconsistency, makes the component easier to use for end users and... it's the only one implementation of PSR-13 for now.

nicolas-grekas commented on src/Symfony/Component/WebLink/EventListener/AddLinkHeaderListener.php:
```diff
@@ -9,26 +9,28 @@
+    public function __construct()
     {
-        $this->preloadManager = $preloadManager;
+        $this->serializer = new HttpHeaderSerializer();
```
having this class be instantiated directly here means it's an internal implementation detail to the listener
imho, we should either make it a service, or merge the implement in this class, don't you think?

dunglas replied:
It's indeed an internal detail of the listener.

But the HttpHeaderSerializer has can be used alone through the component. I don't think it makes sense (for now) to expose it as a service but it may be useful for library creators to serialize a list of Link instances to a valid HTTP header value (it's almost all this component do).

I would keep this as is (btw the serializer is final but not internal for the same reason).

dunglas committed:
Fix @nicolasgrekas' comments

eko commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/web_link.xml:
```diff
@@ -0,0 +1,14 @@
+
+    <services>
+
+        <service id="web_link.add_link_header_listener" class="Symfony\Component\Link\WebLink\AddLinkHeaderListener" public="false">
```
It looks like there is a typo in class name here, should be: `Symfony\Component\WebLink\AddLinkHeaderListener`

eko replied:
It looks like there is a typo in class name here, should be: `Symfony\Component\WebLink\AddLinkHeaderListener`

eko commented on src/Symfony/Component/WebLink/composer.json:
```diff
@@ -0,0 +1,43 @@
+    "name": "symfony/web-link",
+    "type": "library",
+    "description": "Symfony WebLink Component",
+    "keywords": ["link", "psr13", "http", "HTTP/2", "preload", "prefetch", "pre-render", "dns-prefetch", "push", "performance"],
```
I think `pre-render` here should be fixed to `prerender`?

dunglas committed:
Fix @eko's comments

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -18,10 +18,10 @@
 use Symfony\Component\Serializer\Serializer;
 use Symfony\Component\Translation\Translator;
 use Symfony\Component\Validator\Validation;
+use Symfony\Component\WebLink\WebLinkManagerInterface;
```
this interface does not exist anymore

dunglas committed:
Remove references to WebLinkManagerInterface

dunglas committed:
Require fig/link-util in the WebLink component

nicolas-grekas reviewed (approved):
:+1:

fabpot commented:
Some tests are broken on Travis.

dunglas commented:
@fabpot IIUC, only because the component is not available on Packagist yet (DEPS=low)?

fabpot commented:
Thank you @dunglas.

TomasVotruba commented:
Great job! 

I almost missed it... is there a post somewhere? I couldn't find any.  Or is it reverted?

jdreesen commented:
There is [New in Symfony 3.3: WebLink component](https://symfony.com/blog/new-in-symfony-3-3-weblink-component).

TomasVotruba commented:
@jdreesen Ah, thank you. I probably looked for "Link*".
